unit/size_left_test.c

```diff
@@ -0,0 +1,442 @@
+}
+
+/* returns 0 on success or a negative value that can be stringified with
+ * fi_strerror on error */
+static int setup_ep_fixture(struct fid_ep **ep_o)
+{
+	int ret;
+	struct fi_info *myfi;
+	struct fi_av_attr av_attr;
+	struct fi_cq_attr cq_attr;
+
+	assert(ep_o != NULL);
+	ret = 0;
+
+	myfi = fi_dupinfo(fi);
+	if (myfi == NULL) {
+		printf("fi_dupinfo returned NULL\n");
+		goto fail;
+	}
+
+	ret = fi_endpoint(domain, myfi, ep_o, NULL);
+	if (ret != 0) {
+		printf("fi_endpoint %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	memset(&cq_attr, 0, sizeof cq_attr);
+	cq_attr.format = FI_CQ_FORMAT_CONTEXT;
+	cq_attr.wait_obj = FI_WAIT_NONE;
+	cq_attr.size = TX_CQ_DEPTH;
+
+	ret = fi_cq_open(domain, &cq_attr, &wcq, /*context=*/NULL);
+	if (ret != 0) {
+		printf("fi_cq_open %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	memset(&cq_attr, 0, sizeof cq_attr);
+	cq_attr.format = FI_CQ_FORMAT_CONTEXT;
+	cq_attr.wait_obj = FI_WAIT_NONE;
+	cq_attr.size = RX_CQ_DEPTH;
+
+	ret = fi_cq_open(domain, &cq_attr, &rcq, /*context=*/NULL);
+	if (ret != 0) {
+		printf("fi_cq_open %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	memset(&av_attr, 0, sizeof av_attr);
+	av_attr.type = FI_AV_MAP;
+	av_attr.count = 1;
+	av_attr.name = NULL;
+
+	ret = fi_av_open(domain, &av_attr, &av, NULL);
+	if (ret != 0) {
+		printf("fi_av_open %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	ret = fi_ep_bind(*ep_o, &wcq->fid, FI_SEND);
+	if (ret != 0) {
+		printf("fi_ep_bind(wcq) %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	ret = fi_ep_bind(*ep_o, &rcq->fid, FI_RECV);
+	if (ret != 0) {
+		printf("fi_ep_bind(rcq) %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	ret = fi_ep_bind(*ep_o, &av->fid, 0);
+	if (ret != 0) {
+		printf("fi_ep_bind(av) %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	ret = fi_enable(*ep_o);
+	if (ret != 0) {
+		printf("fi_enable %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	if (myfi != NULL) {
+		fi_freeinfo(myfi);
+	}
+	return ret;
+
+fail:
+	if (myfi != NULL) {
+		fi_freeinfo(myfi);
+	}
+	return teardown_ep_fixture(*ep_o);
+}
+
+static int
+rx_size_left(void)
+{
+	int ret;
+	int testret;
+	struct fid_ep *ep;
+
+	testret = FAIL;
+	ep = NULL;
+
+	ret = setup_ep_fixture(&ep);
+	if (ret != 0) {
+		printf("failed to setup test fixture: %s\n", fi_strerror(-ret));
+		goto fail;
+	}
+
+	ret = fi_rx_size_left(ep);
+	if (ret < 0) {
+		printf("fi_rx_size_left returned %d (-%s)\n", ret,
+			fi_strerror(-ret));
+		goto fail;
+	}
+
+	/* TODO: once fi_rx_attr's size field meaning has been fixed to refer to
+	 * queue depth instead of number of bytes, we can do a little basic
+	 * sanity checking here */
+
+	testret = PASS;
+fail:
+	ret = teardown_ep_fixture(ep);
+	if (ret != 0)
+		testret = FAIL;
+	return testret;
+}
+
+static int
+rx_size_left_err(void)
+{
+	int ret;
+	int testret;
+	struct fid_ep *ep;
+	struct fi_info *myfi;
+
+	testret = FAIL;
+	ep = NULL;
+
+	myfi = fi_dupinfo(fi);
+
+	/* datapath operation, not expected to be caught by libfabric */
+#if 0
```

Comment:
Yes.  It's a very straightforward "check for error handling" unit test that I would expect anyone writing a unit test to include, so I thought it made the most sense to leave it and explain why it was deliberately disabled instead of just omitting it.